ContextManager onInitialized callback for 'RULE_ALTERED_JOB_WORKER' failed.

### Which version of ShardingSphere did you use?

`master`

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

`ShardingSphere-Proxy`

```
[ERROR] 2022-03-19 19:39:52.560 [main] o.a.s.p.i.BootstrapInitializer - contextManager onInitialized callback for 'RULE_ALTERED_JOB_WORKER' failed
org.apache.shardingsphere.mode.repository.cluster.ClusterPersistRepositoryException: java.lang.IllegalStateException: Already started
	at org.apache.shardingsphere.mode.repository.cluster.zookeeper.handler.CuratorZookeeperExceptionHandler.handleException(CuratorZookeeperExceptionHandler.java:51)
	at org.apache.shardingsphere.mode.repository.cluster.zookeeper.CuratorZookeeperRepository.start(CuratorZookeeperRepository.java:260)
	at org.apache.shardingsphere.mode.repository.cluster.zookeeper.CuratorZookeeperRepository.watch(CuratorZookeeperRepository.java:251)
	at org.apache.shardingsphere.data.pipeline.core.api.impl.GovernanceRepositoryAPIImpl.watch(GovernanceRepositoryAPIImpl.java:120)
	at org.apache.shardingsphere.data.pipeline.core.execute.PipelineJobExecutor.watchGovernanceRepositoryConfiguration(PipelineJobExecutor.java:52)
	at org.apache.shardingsphere.data.pipeline.core.execute.PipelineJobExecutor.doStart(PipelineJobExecutor.java:48)
	at org.apache.shardingsphere.data.pipeline.api.executor.AbstractLifecycleExecutor.start(AbstractLifecycleExecutor.java:41)
	at org.apache.shardingsphere.data.pipeline.scenario.rulealtered.RuleAlteredJobWorker.initWorkerIfNecessary(RuleAlteredJobWorker.java:117)
	at org.apache.shardingsphere.data.pipeline.scenario.rulealtered.spi.ContextManagerLifecycleListenerImpl.onInitialized(ContextManagerLifecycleListenerImpl.java:46)
	at org.apache.shardingsphere.proxy.initializer.BootstrapInitializer.contextManagerInitializedCallback(BootstrapInitializer.java:80)
	at org.apache.shardingsphere.proxy.initializer.BootstrapInitializer.init(BootstrapInitializer.java:59)
	at org.apache.shardingsphere.proxy.Bootstrap.main(Bootstrap.java:47)
Caused by: java.lang.IllegalStateException: Already started
	at org.apache.curator.shaded.com.google.common.base.Preconditions.checkState(Preconditions.java:507)
	at org.apache.curator.framework.recipes.cache.CuratorCacheImpl.start(CuratorCacheImpl.java:88)
	at org.apache.shardingsphere.mode.repository.cluster.zookeeper.CuratorZookeeperRepository.start(CuratorZookeeperRepository.java:256)
	... 10 common frames omitted
[INFO ] 2022-03-19 19:39:52.610 [main] o.a.s.p.v.ShardingSphereProxyVersion - Database name is `MySQL`, version is `8.0.25`
[INFO ] 2022-03-19 19:39:52.896 [main] o.a.s.p.frontend.ShardingSphereProxy - ShardingSphere-Proxy Cluster mode started successfully
```

### Configuration

```
schemaName: sharding_db

dataSources:
 ds_0:
   url: jdbc:mysql://127.0.0.1:3306/demo_ds_0?serverTimezone=UTC&useSSL=false
   username: root
   password: 123456
   connectionTimeoutMilliseconds: 30000
   idleTimeoutMilliseconds: 60000
   maxLifetimeMilliseconds: 1800000
   maxPoolSize: 50
   minPoolSize: 1
 ds_1:
   url: jdbc:mysql://127.0.0.1:3306/demo_ds_1?serverTimezone=UTC&useSSL=false
   username: root
   password: 123456
   connectionTimeoutMilliseconds: 30000
   idleTimeoutMilliseconds: 60000
   maxLifetimeMilliseconds: 1800000
   maxPoolSize: 50
   minPoolSize: 1

rules:
- !SHARDING
 tables:
   t_order:
     actualDataNodes: ds_${0..1}.t_order_${0..1}
     tableStrategy:
       standard:
         shardingColumn: order_id
         shardingAlgorithmName: t_order_inline
     keyGenerateStrategy:
       column: order_id
       keyGeneratorName: snowflake
   t_order_item:
     actualDataNodes: ds_${0..1}.t_order_item_${0..1}
     tableStrategy:
       standard:
         shardingColumn: order_id
         shardingAlgorithmName: t_order_item_inline
     keyGenerateStrategy:
       column: order_item_id
       keyGeneratorName: snowflake
 bindingTables:
   - t_order,t_order_item
 defaultDatabaseStrategy:
   standard:
     shardingColumn: user_id
     shardingAlgorithmName: database_inline
 defaultTableStrategy:
   none:
 
 shardingAlgorithms:
   database_inline:
     type: INLINE
     props:
       algorithm-expression: ds_${user_id % 2}
   t_order_inline:
     type: INLINE
     props:
       algorithm-expression: t_order_${order_id % 2}
   t_order_item_inline:
     type: INLINE
     props:
       algorithm-expression: t_order_item_${order_id % 2}
 
 keyGenerators:
   snowflake:
     type: SNOWFLAKE
```

--- a/shardingsphere-mode/shardingsphere-mode-core/src/main/java/org/apache/shardingsphere/mode/metadata/persist/node/ComputeNode.java
+++ b/shardingsphere-mode/shardingsphere-mode-core/src/main/java/org/apache/shardingsphere/mode/metadata/persist/node/ComputeNode.java
@@ -105,6 +105,15 @@ public final class ComputeNode {
     }
     
     /**
+     * Get compute node path.
+     *
+     * @return compute node path
+     */
+    public static String getComputeNodePath() {
+        return String.join("/", "", ROOT_NODE, COMPUTE_NODE);
+    }
+    
+    /**
      * Get instance worker id node path.
      *
      * @param instanceId instance id

--- a/shardingsphere-mode/shardingsphere-mode-core/src/test/java/org/apache/shardingsphere/mode/metadata/persist/node/ComputeNodeTest.java
+++ b/shardingsphere-mode/shardingsphere-mode-core/src/test/java/org/apache/shardingsphere/mode/metadata/persist/node/ComputeNodeTest.java
@@ -88,4 +88,9 @@ public final class ComputeNodeTest {
         assertThat(actual.get().getInstanceId().getId(), is("127.0.0.1@3307"));
         assertThat(actual.get().getInstanceType(), is(InstanceType.JDBC));
     }
+    
+    @Test
+    public void assertGetComputeNodePath() {
+        assertThat(ComputeNode.getComputeNodePath(), is("/nodes/compute_nodes"));
+    }
 }

--- a/shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/GovernanceWatcherFactory.java
+++ b/shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/GovernanceWatcherFactory.java
@@ -22,8 +22,6 @@ import org.apache.shardingsphere.infra.eventbus.ShardingSphereEventBus;
 import org.apache.shardingsphere.mode.repository.cluster.ClusterPersistRepository;
 import org.apache.shardingsphere.spi.ShardingSphereServiceLoader;
 
-import java.util.Collection;
-
 /**
  * Governance watcher factory.
  */
@@ -40,8 +38,22 @@ public final class GovernanceWatcherFactory {
      * Watch listeners.
      */
     public void watchListeners() {
-        Collection<GovernanceWatcher> governanceWatchers = ShardingSphereServiceLoader.getSingletonServiceInstances(GovernanceWatcher.class);
-        repository.watch("/", dataChangedEventListener -> governanceWatchers.stream().filter(each -> each.getWatchingTypes().contains(dataChangedEventListener.getType()))
-                .forEach(each -> each.createGovernanceEvent(dataChangedEventListener).ifPresent(ShardingSphereEventBus.getInstance()::post)));
+        for (GovernanceWatcher<?> each : ShardingSphereServiceLoader.getSingletonServiceInstances(GovernanceWatcher.class)) {
+            watch(each);
+        }
+    }
+    
+    private void watch(final GovernanceWatcher<?> listener) {
+        for (String each : listener.getWatchingKeys()) {
+            watch(each, listener);
+        }
+    }
+    
+    private void watch(final String watchingKey, final GovernanceWatcher<?> listener) {
+        repository.watch(watchingKey, dataChangedEventListener -> {
+            if (listener.getWatchingTypes().contains(dataChangedEventListener.getType())) {
+                listener.createGovernanceEvent(dataChangedEventListener).ifPresent(ShardingSphereEventBus.getInstance()::post);
+            }
+        });
     }
 }

--- a/shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/status/compute/watcher/ComputeNodeStateChangedWatcher.java
+++ b/shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/status/compute/watcher/ComputeNodeStateChangedWatcher.java
@@ -45,7 +45,7 @@ public final class ComputeNodeStateChangedWatcher implements GovernanceWatcher<G
     
     @Override
     public Collection<String> getWatchingKeys() {
-        return Collections.singleton(ComputeNode.getAttributesNodePath());
+        return Collections.singleton(ComputeNode.getComputeNodePath());
     }
     
     @Override

--- a/shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-repository/shardingsphere-cluster-mode-repository-provider/shardingsphere-cluster-mode-repository-zookeeper-curator/src/main/java/org/apache/shardingsphere/mode/repository/cluster/zookeeper/CuratorZookeeperRepository.java
+++ b/shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-repository/shardingsphere-cluster-mode-repository-provider/shardingsphere-cluster-mode-repository-zookeeper-curator/src/main/java/org/apache/shardingsphere/mode/repository/cluster/zookeeper/CuratorZookeeperRepository.java
@@ -49,10 +49,10 @@ import org.apache.zookeeper.data.ACL;
 import java.nio.charset.StandardCharsets;
 import java.util.Collections;
 import java.util.Comparator;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
-import java.util.Optional;
 import java.util.Properties;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.TimeUnit;
@@ -62,7 +62,7 @@ import java.util.concurrent.TimeUnit;
  */
 public final class CuratorZookeeperRepository implements ClusterPersistRepository {
     
-    private CuratorCache curatorCache;
+    private final Map<String, CuratorCache> caches = new HashMap<>();
     
     private CuratorFramework client;
     
@@ -236,8 +236,10 @@ public final class CuratorZookeeperRepository implements ClusterPersistRepositor
     
     @Override
     public void watch(final String key, final DataChangedEventListener listener) {
-        if (!Optional.ofNullable(curatorCache).isPresent()) {
-            curatorCache = CuratorCache.build(client, "/");
+        if (!caches.containsKey(key)) {
+            CuratorCache curatorCache = CuratorCache.build(client, key);
+            start(curatorCache);
+            caches.put(key, curatorCache);
         }
         CuratorCacheListener curatorCacheListener = CuratorCacheListener.builder()
                 .forTreeCache(client, (framework, treeCacheListener) -> {
@@ -247,8 +249,7 @@ public final class CuratorZookeeperRepository implements ClusterPersistRepositor
                                 new String(treeCacheListener.getData().getData(), StandardCharsets.UTF_8), changedType));
                     }
                 }).build();
-        curatorCache.listenable().addListener(curatorCacheListener);
-        start(curatorCache);
+        caches.get(key).listenable().addListener(curatorCacheListener);
     }
     
     private void start(final CuratorCache cache) {
@@ -314,7 +315,7 @@ public final class CuratorZookeeperRepository implements ClusterPersistRepositor
     
     @Override
     public void close() {
-        curatorCache.close();
+        caches.values().forEach(CuratorCache::close);
         waitForCacheClose();
         CloseableUtils.closeQuietly(client);
     }

--- a/shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-repository/shardingsphere-cluster-mode-repository-provider/shardingsphere-cluster-mode-repository-zookeeper-curator/src/test/java/org/apache/shardingsphere/mode/repository/cluster/zookeeper/CuratorZookeeperRepositoryTest.java
+++ b/shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-repository/shardingsphere-cluster-mode-repository-provider/shardingsphere-cluster-mode-repository-zookeeper-curator/src/test/java/org/apache/shardingsphere/mode/repository/cluster/zookeeper/CuratorZookeeperRepositoryTest.java
@@ -53,7 +53,9 @@ import java.lang.reflect.Field;
 import java.nio.charset.StandardCharsets;
 import java.util.Arrays;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 import java.util.Properties;
 import java.util.concurrent.TimeUnit;
 
@@ -203,7 +205,7 @@ public final class CuratorZookeeperRepositoryTest {
     @Test
     @SneakyThrows
     public void assertWatchUpdatedChangedType() {
-        mockCache();
+        mockCache("/test/children_updated/1");
         ChildData oldData = new ChildData("/test/children_updated/1", null, "value1".getBytes());
         ChildData data = new ChildData("/test/children_updated/1", null, "value2".getBytes());
         doAnswer(AdditionalAnswers.answerVoid(getListenerAnswer(CuratorCacheListener.Type.NODE_CHANGED, oldData, data))).when(listenable).addListener(any(CuratorCacheListener.class));
@@ -218,7 +220,7 @@ public final class CuratorZookeeperRepositoryTest {
     
     @Test
     public void assertWatchDeletedChangedType() throws Exception {
-        mockCache();
+        mockCache("/test/children_deleted/5");
         ChildData oldData = new ChildData("/test/children_deleted/5", null, "value5".getBytes());
         ChildData data = new ChildData("/test/children_deleted/5", null, "value5".getBytes());
         doAnswer(AdditionalAnswers.answerVoid(getListenerAnswer(CuratorCacheListener.Type.NODE_DELETED, oldData, data))).when(listenable).addListener(any(CuratorCacheListener.class));
@@ -234,7 +236,7 @@ public final class CuratorZookeeperRepositoryTest {
     @Test
     @SneakyThrows
     public void assertWatchAddedChangedType() {
-        mockCache();
+        mockCache("/test/children_added/4");
         ChildData data = new ChildData("/test/children_added/4", null, "value4".getBytes());
         doAnswer(AdditionalAnswers.answerVoid(getListenerAnswer(CuratorCacheListener.Type.NODE_CREATED, null, data))).when(listenable).addListener(any(CuratorCacheListener.class));
         SettableFuture<DataChangedEvent> settableFuture = SettableFuture.create();
@@ -246,10 +248,12 @@ public final class CuratorZookeeperRepositoryTest {
         assertThat(dataChangedEvent.getValue(), is("value4"));
     }
     
-    private void mockCache() throws Exception {
-        Field cachesFiled = CuratorZookeeperRepository.class.getDeclaredField("curatorCache");
+    private void mockCache(final String key) throws Exception {
+        Field cachesFiled = CuratorZookeeperRepository.class.getDeclaredField("caches");
         cachesFiled.setAccessible(true);
-        cachesFiled.set(REPOSITORY, curatorCache);
+        Map<String, CuratorCache> caches = new HashMap<>();
+        caches.put(key, curatorCache);
+        cachesFiled.set(REPOSITORY, caches);
         when(curatorCache.listenable()).thenReturn(listenable);
     }
     

